File: python/lineSweep.py
# generatl technique used to solve gemoetric problems by sweeping a vertical
# line across the plane 

def line_sweep(segments):
  """
  performs a line sweep algorithm on a set of line segments
  args:
    segments: list of line segments, each represented as a tuple of two points 
  returns:
    list of intersection points
  """
  event_queue = []
  for segment in segments:
    p1, p2 = segment
    event_queue.append((min(p1[0], p2[0]), 'start', segment))
    event_queue.append((max(p1[0], p2[0]), 'end', segment))

  event_queue.sort()
  active_segments = []
  intersections = []

  for event in event_queue:
    x, event_type, segment = event

    if event_type == 'start':
      active_segments.append(segment)
      for other_segment in active_segments:
        if segments_intersect(segment, other_segment):
          intersections.append(intersection_point(segment, other_segment))

    else:
      active_segments.remove(segment)

  return intersections 

File: java/LineSweep.java
// fundamental technique in computational geometry used to solve various problems like intersection detection,
// finding closest pairs, and more; involves sweeping a vertical line across the plane, processing events (line
// segment endpoints) as they are encountered


public class LineSweep
  {
    private PriorityQueue<EventPoint> eventQueue;
    private SweepLineStatus sweepLineStatus;

    public LineSweep(List<LineSegment> lineSegments)
    {
      // initialize event queue and sweep line status
      eventQueue = new PriorityQueue<>(Comparator.comparingDouble(p->p.x));
      sweepLineStatus = new SweepLineStatus();

      // populate event queue with left and right endpoints of line segments
      for(LineSegment segment: lineSegments)
        {
          eventQueue.offer(new EventPoint(segment.p1.x, segment.p1.y, true, segment));
          eventQueue.offer(new EventPoint(segment.p2.x, segment.p2.y, false, segment)):
        }
    }
    public void processEvents()
    {
      while(!eventQueue.isEmpty())
        {
          EventPoint event = eventQueue.poll();
          if(event.isLeftEndpoint)
          {
            sweepLineStatus.insert(event.lineSegment);
          }
          else
          {
            sweepLineStatus.delete(event.lineSegment);
          }
        }
    }
  }

    
